components/script/dom/vrdisplay.rs

```diff
@@ -0,0 +1,618 @@
+            }
+        }
+    }
+
+    // https://w3c.github.io/webvr/#dom-vrdisplay-depthnear
+    fn DepthNear(&self) -> Finite<f64> {
+        Finite::wrap(self.depth_near.get())
+    }
+
+    // https://w3c.github.io/webvr/#dom-vrdisplay-depthnear
+    fn SetDepthNear(&self, value: Finite<f64>) -> () {
+        self.depth_near.set(*value.deref());
+    }
+
+    // https://w3c.github.io/webvr/#dom-vrdisplay-depthfar
+    fn DepthFar(&self) -> Finite<f64> {
+        Finite::wrap(self.depth_far.get())
+    }
+
+    // https://w3c.github.io/webvr/#dom-vrdisplay-depthfar
+    fn SetDepthFar(&self, value: Finite<f64>) -> () {
+        self.depth_far.set(*value.deref());
+    }
+
+    // https://w3c.github.io/webvr/#dom-vrdisplay-requestanimationframe
+    fn RequestAnimationFrame(&self, callback: Rc<FrameRequestCallback>) -> u32 {
+        if self.presenting.get() {
+            let raf_id = self.next_raf_id.get();
+            self.next_raf_id.set(raf_id + 1);
+            self.raf_callback_list.borrow_mut().push((raf_id, Some(callback)));
+            raf_id
+        } else {
+            // WebVR spec: When a VRDisplay is not presenting it should
+            // fallback to window.requestAnimationFrame.
+            self.global().as_window().RequestAnimationFrame(callback)
+        }
+    }
+
+    // https://w3c.github.io/webvr/#dom-vrdisplay-cancelanimationframe
+    fn CancelAnimationFrame(&self, handle: u32) -> () {
+        if self.presenting.get() {
+            let mut list = self.raf_callback_list.borrow_mut();
+            if let Some(mut pair) = list.iter_mut().find(|pair| pair.0 == handle) {
+                pair.1 = None;
+            }
+        } else {
+            // WebVR spec: When a VRDisplay is not presenting it should
+            // fallback to window.cancelAnimationFrame.
+            self.global().as_window().CancelAnimationFrame(handle);
+        }
+    }
+
+    #[allow(unrooted_must_root)]
+    // https://w3c.github.io/webvr/#dom-vrdisplay-requestpresent
+    fn RequestPresent(&self, layers: Vec<VRLayer>) -> Rc<Promise> {
+        let promise = Promise::new(&self.global());
+        // TODO: WebVR spec: this method must be called in response to a user gesture
+
+        // WebVR spec: If canPresent is false the promise MUST be rejected
+        if !self.display.borrow().0.capabilities.can_present {
+            let msg = "VRDisplay canPresent is false".to_string();
+            promise.reject_native(promise.global().get_cx(), &msg);
+            return promise;
+        }
+
+        // Current WebVRSpec only allows 1 VRLayer if the VRDevice can present.
+        // Future revisions of this spec may allow multiple layers to enable more complex rendering effects
+        // such as compositing WebGL and DOM elements together.
+        // That functionality is not allowed by this revision of the spec.
+        if layers.len() != 1 {
+            let msg = "The number of layers must be 1".to_string();
```

Comment:
I had to use string because of this: 
```
283 |             promise.reject_native(promise.global().get_cx(), msg);
    |                     ^^^^^^^^^^^^^ the trait `core::marker::Sized` is not implemented for `str`
```